src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,494 @@
+--     apply ideal.mem_Sup_of_mem, use I.2,
+--     simp only [subtype.val_eq_coe] at h𝓈 ⊢, rw h𝓈,
+--     refine ideal.subset_span _, rw [mem_image], use x', refine ⟨hx1, hx2⟩, },
+--   intros I, exact I.2,
+-- end
+
+-- variables (A)
+
+-- instance : has_inf (homogeneous_ideal A) :=
+-- { inf := λ I J, ⟨I ⊓ J, I.prop.inf J.prop⟩ }
+
+-- instance : has_Inf (homogeneous_ideal A) :=
+-- { Inf := λ ℐ, ⟨Inf (coe '' ℐ), homogeneous_ideal.Inf $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+-- instance : has_sup (homogeneous_ideal A) :=
+-- { sup := λ I J, ⟨I ⊔ J, I.prop.sup J.prop⟩ }
+
+-- instance : has_Sup (homogeneous_ideal A) :=
+-- { Sup := λ ℐ, ⟨Sup (coe '' ℐ), ideal.is_homogeneous.Sup $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+-- instance : has_mul (homogeneous_ideal A) :=
+-- { mul := λ I J, ⟨I * J, I.prop.mul J.prop⟩ }
+
+-- instance : has_add (homogeneous_ideal A) := ⟨(⊔)⟩
+
+end operations
+
+section homogeneous_core
+
+open set_like direct_sum set
+
+variables {ι R : Type*} [comm_ring R]
+variables (A : ι → ideal R)
+variable (I : ideal R)
+
+/-- For any `I : ideal R`, not necessarily homogeneous, there is a homogeneous ideal associated with
+`I` spanned by all homogeneous elements in `I`. This construction is used when proving that the
+radical of a homogeneous ideal is homogeneous. -/
+def ideal.homogeneous_core : ideal R := ideal.span (set_of (is_homogeneous A) ∩ I)
+
+lemma ideal.homogeneous_core_is_mono : monotone (ideal.homogeneous_core A) := λ I J I_le_J,
+begin
+  apply ideal.span_mono, rintros r ⟨hr1, hr2⟩,
+  refine ⟨hr1, I_le_J hr2⟩,
+end
+
+lemma ideal.homogeneous_core_le_ideal :
+  ideal.homogeneous_core A I ≤ I :=
+begin
+  rw ideal.homogeneous_core,
+  conv_rhs { rw ←ideal.span_eq I },
+  apply ideal.span_mono, exact (set_of (is_homogeneous A)).inter_subset_right ↑I,
+end
+
+variables [add_comm_monoid ι] [decidable_eq ι]
+variables [graded_algebra A]
+
+lemma ideal.is_homogeneous.homogeneous_core :
+  ideal.is_homogeneous A (ideal.homogeneous_core A I) :=
+begin
+  rw ideal.is_homogeneous.iff_exists,
+  use ({x | ↑x ∈ I}),
+  rw ideal.homogeneous_core, congr, ext, split; intro hx;
+  simp only [mem_inter_eq, mem_set_of_eq, set_like.mem_coe] at hx ⊢,
+  use x, exact hx.1, refine ⟨hx.2, rfl⟩,
+  obtain ⟨y, hy₁, hy₂⟩ := hx, simp only [mem_set_of_eq] at hy₁, split, rw ←hy₂,
+  rcases y with ⟨y, ⟨i, hy₃⟩⟩, use i, refine hy₃,
+  rw ←hy₂, refine hy₁,
+end
+
+lemma ideal.is_homogeneous.homogeneous_core_eq_self [Π (i : ι) (x : A i), decidable (x ≠ 0)]
+  (h : ideal.is_homogeneous A I) :
+  ideal.homogeneous_core A I = I :=
+begin
+  ext x, split; intros hx,
+  { apply ideal.homogeneous_core_le_ideal, exact hx, },
+  { rw ←graded_algebra.sum_support_decompose A x,
+    refine ideal.sum_mem _ _,
+    intros i hi, apply ideal.subset_span, split,
+    use i, exact submodule.coe_mem _, apply h, exact hx },
+end
+
+-- lemma ideal.homogeneous_core.eq_Sup [Π (i : ι) (x : A i), decidable (x ≠ 0)] :
```

Comment:
This is commented out because the proof actually needs that `Sup` of homogeneous ideal is homogeneous. So I comment it out to avoid circular definition.